includes/utilities.inc

```diff
@@ -20,12 +20,15 @@ function islandora_usage_stats_log_object_view(AbstractObject $object) {
       $pid_id = islandora_usage_stats_get_pid_id($object->id);
       $object_access_id = islandora_usage_stats_object_view_to_db($pid_id);
 
+      module_invoke_all('islandora_usage_stats_access_recorded', 'object', $object_access_id, $pid_id, $object->id);
```

Comment:
Folks: @mjordan @bryjbrown wonder if making this new hook less generic could be something you would agree on? ```islandora_usage_stats_access_recorded``` right now  requires some conditionals (in their implementation) based on $type. That actually requires some extra care for implementers (i'm not one but I would assume its easy to avoid checking for all cases) like only taking care if $pid if passed.. and of type object, or collection, etc.. So what about A) splitting "object based" and "DS based" into 2 different ones, and passing $object instead of the pid for the first one. that means, if someone wants to record the models, well, it can be extracted from there but no need to invoke twice. Also, right now types are restricting something that actually does not need to be restricted at that level. What if someone wants to record a book access if a page was being seen? To avoid taking all "types" i account, maybe that could be left to the implementer by just passing the $object itself.

Reply:
@DiegoPino thanks for comments.

> for sake of future documentation, could you add the use cases to the pull request message

Use cases are mentioned in the JIRA ticket, e.g. "By implementing this hook, modules could persist additional data such as referrer, or geolocation data for the event." I like the idea of supplying use case, but I wonder if the JIRA ticket is good enough for that.

> islandora_usage_stats_access_recorded right now requires some conditionals (in their implementation) based on $type.

hook_islandora_object_alter, which uses a $context that has numerous options within each sub-array. How is that any different than the hook definition in this PR? 

> Also, right now types are restricting something that actually does not need to be restricted at that level.

Those four types are the only ones that he Stats base module supports. I probably missing something, but how would not following that pattern in this hook definition get around that? I agree that passing the `$object` is always useful though, simply because you can do pretty much anything in the implementation.

Reply:
@mjordan thanks. So Sorry i missed the use case from the email (damn phones and their tiny screens)!
i get your point about hook_islandora_object_alter and *in no way i'm blocking this*,

I just feel there are too many invocations to the same set of hooks and for that, I felt 

A) having different hooks would make more sense, or, 

B) invoke them only once and actually pass a context. 

Because that hook(islandora_object_alter) is invoked a single time for a single PHP object type,
 In this case, the invocation is multiple in a single run and operating over two different (possible?) types of PHP Objects, `datastreams` and `Objects` letting (forcing) the implementer to discern when to actually do heavy work.

Sometimes it can be good to allow a hook  to be a little less smart because it will be only invoked under certain circumstances, imagine like prefiltered reality. I only say this because, yes, we have pretty smart hooks that deal with the full rainbow of scenarios in islandora too. So not a  must.

What i'm trying to say here is that your hook implementation will have to, at least, have something like this to work properly
```PHP
if (isset($pid) && ($type=='object)) { 
// deal with objects, but not cmodels which are objects, and  also not collections which are also objects) 
}
else { 
  if ($type!= `ds`) {
   //do datastream stuff 
  } 
  else {
    // do collection, cmodel stuff
   }
}
```

So basically that becomes a template. If that is the case and you all are OK with that, maybe adding that to the API hook body (the example) to let people understand what happens and what is needed to have a sane experience writing this hook could be a plus. Thanks!

Reply:
I'm OK with it, but also interested to hear what others say. I agree that the example code you provide here would be a useful addition to the `.api.php` file and am happy to add it.

Reply:
I think @DiegoPino has convinced me here, too. Having separate hooks fire for when objects are recorded vs datastreams (and perhaps even breaking it out further so that you can implement a hook that only fires when the MODS is viewed for example) would give implementers more options.

Reply:
OK, let me propose something, in a little while.